vcert/connection_tpp.py

```diff
@@ -109,11 +129,220 @@ def auth(self):
+        if not request.id and not request.thumbprint:
+            log.debug("Request id or thumbprint must be specified for TPP")
+            raise CertificateRenewError
+        if not request.id and request.thumbprint:
+            request.id = self.search_by_thumbprint(request.thumbprint)
+        if reuse_key:
+            log.debug("Trying to renew certificate %s" % request.id)
+            status, data = self._post(URLS.CERTIFICATE_RENEW, data={"CertificateDN": request.id})
+            if not data['Success']:
+                raise CertificateRenewError
+            return
+        cert = self.retrieve_cert(request)
+        cert = x509.load_pem_x509_certificate(cert.cert.encode(), default_backend())
+        for a in cert.subject:
+            if a.oid == x509.NameOID.COMMON_NAME:
+                request.common_name = a.value
+            elif a.oid == x509.NameOID.COUNTRY_NAME:
+                request.country = a.value
+            elif a.oid == x509.NameOID.LOCALITY_NAME:
+                request.locality = a.value
+            elif a.oid == x509.NameOID.STATE_OR_PROVINCE_NAME:
+                request.province = a.value
+            elif a.oid == x509.NameOID.ORGANIZATION_NAME:
+                request.organization = a.value
+            elif a.oid == x509.NameOID.ORGANIZATIONAL_UNIT_NAME:
+                request.organizational_unit = a.value
+        for e in cert.extensions:
+            if e.oid == x509.OID_SUBJECT_ALTERNATIVE_NAME:
+                request.san_dns = list([x.value for x in e.value if isinstance(x, x509.DNSName)])
+                request.email_addresses = list([x.value for x in e.value if isinstance(x, x509.RFC822Name)])
+                request.ip_addresses = list([x.value.exploded for x in e.value if isinstance(x, x509.IPAddress)])
+        if cert.signature_algorithm_oid in (algos.ECDSA_WITH_SHA1, algos.ECDSA_WITH_SHA224, algos.ECDSA_WITH_SHA256,
+                                            algos.ECDSA_WITH_SHA384, algos.ECDSA_WITH_SHA512):
+            request.key_type = (KeyType.ECDSA, KeyType.ALLOWED_CURVES[0])
+        else:
+            request.key_type = KeyType(KeyType.RSA, 2048)  # todo: make parsing key size
+        if not request.csr:
+            request.build_csr()
+        status, data = self._post(URLS.CERTIFICATE_RENEW,
+                                  data={"CertificateDN": request.id, "PKCS10": request.csr})
+        if status == HTTPStatus.OK:
+            if "CertificateDN" in data:
+                request.id = data['CertificateDN']
+            log.debug("Certificate successfully requested with request id %s." % request.id)
+            return True
+        else:
+            log.error("Request status is not %s. %s." % HTTPStatus.OK, status)
+            raise CertificateRequestError
+
+    @staticmethod
+    def _parse_zone_config_to_policy(data):
+        # todo: parse over values to regexps (dont forget tests!)
+        p = data["Policy"]
```

Comment:
How safe is it in this method to assume that these dictionary keys resolve? I'm new to this code, but I usually think thrice before trying to access a node in the dictionary without .get().

Reply:
Thanks for this suggestion as well. Although the code itself was not added in this PR (it was present before), it is nice to have potentially dangerous code design choices pinpointed